### PR TITLE
Валидация элемента РМ

### PR DESCRIPTION
**Pull request by xakep139**
This change is [Reviewable](https://reviewable.io/reviews/2gis/nuclear-vstore/2)

### CODE REVIEW

**Comment by denisivan0v** [comment]
Reviewed 42 of 42 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[VStore.Host/Controllers/ObjectController.cs, line 159 at r1](https://reviewable.io:443/reviews/2gis/nuclear-vstore/2#-KYWWtEHuoMw1y-YCYje:-KYWWtEI7dD7s2FHvS9U:bappf9z) ([raw file](https://github.com/2gis/nuclear-vstore/blob/c01e1980d4553cd5312e8b0ab2c0007ad18c686f/VStore.Host/Controllers/ObjectController.cs#L159)):*
> "type"
Константы давай вынесем в статиский класс какой-нибудь. Есть уже класс Tokens

---

*[VStore/Objects/Validate/ElementFormattedTextTagNames.cs, line 1 at r1](https://reviewable.io:443/reviews/2gis/nuclear-vstore/2#-KYWYWd0ZfTgjBycZxSy:-KYWYWd0ZfTgjBycZxSz:bdnp7wj) ([raw file](https://github.com/2gis/nuclear-vstore/blob/c01e1980d4553cd5312e8b0ab2c0007ad18c686f/VStore/Objects/Validate/ElementFormattedTextTagNames.cs#L1)):*
> NuClear.VStore.Objects.Validate
NuClear.VStore.Objects.ContentValidation

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/2gis/nuclear-vstore/2)*

**Comment by xakep139** [comment]
Review status: 10 of 43 files reviewed at latest revision, 2 unresolved discussions.

---

*[VStore.Host/Controllers/ObjectController.cs, line 159 at r1](https://reviewable.io:443/reviews/2gis/nuclear-vstore/2#-KYWWtEHuoMw1y-YCYje:-KYWj6D2r7p8COWiTM7h:b-896fix) ([raw file](https://github.com/2gis/nuclear-vstore/blob/c01e1980d4553cd5312e8b0ab2c0007ad18c686f/VStore.Host/Controllers/ObjectController.cs#L159)):*
<details><summary><i>Previously, denisivan0v (Denis Ivanov) wrote…</i></summary><blockquote>

> "type"
Константы давай вынесем в статиский класс какой-нибудь. Есть уже класс Tokens
</blockquote></details>

Done.

---

*[VStore/Objects/ContentValidation/ElementFormattedTextTagNames.cs, line 1 at r1](https://reviewable.io:443/reviews/2gis/nuclear-vstore/2#-KYWYWd0ZfTgjBycZxSy:-KYWj8A3bG1bV7PXsE44:b-896fix) ([raw file](https://github.com/2gis/nuclear-vstore/blob/c01e1980d4553cd5312e8b0ab2c0007ad18c686f/VStore/Objects/ContentValidation/ElementFormattedTextTagNames.cs#L1)):*
<details><summary><i>Previously, denisivan0v (Denis Ivanov) wrote…</i></summary><blockquote>

> NuClear.VStore.Objects.Validate
NuClear.VStore.Objects.ContentValidation
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/2gis/nuclear-vstore/2)*

File: VStore.Host/Controllers/ObjectController.cs [hunk 1]
```diff
@@ -86,11 +98,131 @@ public ObjectController(ObjectStorageReader objectStorageReader, ObjectManagemen
...
+            catch (AggregateException exc)
+            {
+                return new JsonResult(GenerateErrorJsonResult(exc))
+                {
+                    StatusCode = 422
+                };
+            }
+            catch (InvalidOperationException exc)
+            {
+                return BadRequest(exc.Message);
```
Comment (xakep139) [inline VStore.Host/Controllers/ObjectController.cs]: Может быть, не стоит сообщать об ошибке, а просто её логировать

**Comment by xakep139** [comment]
Review status: 9 of 52 files reviewed at latest revision, 2 unresolved discussions.

---

*[VStore.Host/Controllers/ObjectController.cs, line 123 at r2](https://reviewable.io:443/reviews/2gis/nuclear-vstore/2#-KYWVrRqdUacY5BwiOlb-r2-123:-KYmXjPYyERi3Fcoxy32:b-896fix) ([raw file](https://github.com/2gis/nuclear-vstore/blob/a5a3b7111fbb93e9293e45b58cc36748bb3f60c2/VStore.Host/Controllers/ObjectController.cs#L123)):*
<details><summary><i>Previously, xakep139 (Nikita Balabaev) wrote…</i></summary><blockquote>

Может быть, не стоит сообщать об ошибке, а просто её логировать
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/2gis/nuclear-vstore/2)*

**Comment by denisivan0v** [comment]
Reviewed 50 of 65 files at r2, 17 of 21 files at r3, 13 of 13 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[VStore.Host/Controllers/ObjectController.cs, line 121 at r3](https://reviewable.io:443/reviews/2gis/nuclear-vstore/2#-KYwikI3TZX8wB6qFGth:-KYwikI3TZX8wB6qFGti:boi4uf7) ([raw file](https://github.com/2gis/nuclear-vstore/blob/f0a2d89b33a5cd9891690fe12dd384e60ee2b1c3/VStore.Host/Controllers/ObjectController.cs#L121)):*
> exc

Давай просто `ex`

---

*[VStore.Host/Controllers/ObjectController.cs, line 129 at r4](https://reviewable.io:443/reviews/2gis/nuclear-vstore/2#-KZ0J79_BwzFCEPorGPQ:-KZ0J79_BwzFCEPorGPR:b-2qplx5) ([raw file](https://github.com/2gis/nuclear-vstore/blob/4700ab6c424cfb3894ecbbbbe343c4e535c35147/VStore.Host/Controllers/ObjectController.cs#L129)):*
> new EventId(0)

0 - это нормально?

---

*[VStore.Host/Controllers/ObjectController.cs, line 228 at r4](https://reviewable.io:443/reviews/2gis/nuclear-vstore/2#-KZ0JHVQxzsYR3jY740s:-KZ0JHVQxzsYR3jY740t:b-98sap8) ([raw file](https://github.com/2gis/nuclear-vstore/blob/4700ab6c424cfb3894ecbbbbe343c4e535c35147/VStore.Host/Controllers/ObjectController.cs#L228)):*
> char.ToLower

В Startup.cs есть настройки JSON.NET, там включен camelCase

---

*[VStore/Descriptors/Objects/DateElementValue.cs, line 5 at r4](https://reviewable.io:443/reviews/2gis/nuclear-vstore/2#-KZ0KdXzoobhPKV7Zhw-:-KZ0KdXzoobhPKV7Zhw0:bczctm8) ([raw file](https://github.com/2gis/nuclear-vstore/blob/4700ab6c424cfb3894ecbbbbe343c4e535c35147/VStore/Descriptors/Objects/DateElementValue.cs#L5)):*
> ```Smalltalk
> namespace NuClear.VStore.Descriptors.Objects
> {
>     public class DateElementValue : IObjectElementValue
> ```

Тут есть противоречие. Raw исключает BeginDate и EndDate, и наоборот.  Тип `DateElementValue` используется для сериализации, поэтому BeginDate и EndDate окажутся в формате.

Давай либо 
- в Raw хранить интервал в каком-то формате, например `2016-12-01T00:00:00.000Z;2016-12-31T23:59:59.000Z`, и заведем специальный класс, который будет уметь парсить Raw и который мы будем использовать в проверках
- либо вообще откажемся от Raw в этом случае

---

*[VStore/Objects/ObjectManagementService.cs, line 46 at r4](https://reviewable.io:443/reviews/2gis/nuclear-vstore/2#-KZ0VIXEZlYOK5ykkd0a:-KZ0VIXEZlYOK5ykkd0b:bz1g5sf) ([raw file](https://github.com/2gis/nuclear-vstore/blob/4700ab6c424cfb3894ecbbbbe343c4e535c35147/VStore/Objects/ObjectManagementService.cs#L46)):*
> ValidationRule

В будущем наверно стоит подумать об объектной модели валидации, где логику инкапсулировать в классах, а не процедурах

---

*[VStore/Objects/ObjectManagementService.cs, line 92 at r4](https://reviewable.io:443/reviews/2gis/nuclear-vstore/2#-KZ0MhSz4T5axjHWtoQB:-KZ0MhSz4T5axjHWtoQC:bgh21ay) ([raw file](https://github.com/2gis/nuclear-vstore/blob/4700ab6c424cfb3894ecbbbbe343c4e535c35147/VStore/Objects/ObjectManagementService.cs#L92)):*
> IsFormatted

FasComment не может быть форматированным

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/2gis/nuclear-vstore/2)*

**Comment by xakep139** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[VStore.Host/Controllers/ObjectController.cs, line 121 at r3](https://reviewable.io:443/reviews/2gis/nuclear-vstore/2#-KYwikI3TZX8wB6qFGth:-KZ0YBgacTsHAHV3VVkB:b2o2gal) ([raw file](https://github.com/2gis/nuclear-vstore/blob/f0a2d89b33a5cd9891690fe12dd384e60ee2b1c3/VStore.Host/Controllers/ObjectController.cs#L121)):*
<details><summary><i>Previously, denisivan0v (Denis Ivanov) wrote…</i></summary><blockquote>

> exc

Давай просто `ex`
</blockquote></details>

Уже

---

*[VStore.Host/Controllers/ObjectController.cs, line 129 at r4](https://reviewable.io:443/reviews/2gis/nuclear-vstore/2#-KZ0J79_BwzFCEPorGPQ:-KZ0d6F7ORr8FOP6Svid:b33jvl) ([raw file](https://github.com/2gis/nuclear-vstore/blob/4700ab6c424cfb3894ecbbbbe343c4e535c35147/VStore.Host/Controllers/ObjectController.cs#L129)):*
<details><summary><i>Previously, denisivan0v (Denis Ivanov) wrote…</i></summary><blockquote>

> new EventId(0)

0 - это нормально?
</blockquote></details>

Да

---

*[VStore.Host/Controllers/ObjectController.cs, line 228 at r4](https://reviewable.io:443/reviews/2gis/nuclear-vstore/2#-KZ0JHVQxzsYR3jY740s:-KZ0cxB1hRR8fp4nMS9Q:b-g3ox65) ([raw file](https://github.com/2gis/nuclear-vstore/blob/4700ab6c424cfb3894ecbbbbe343c4e535c35147/VStore.Host/Controllers/ObjectController.cs#L228)):*
<details><summary><i>Previously, denisivan0v (Denis Ivanov) wrote…</i></summary><blockquote>

> char.ToLower

В Startup.cs есть настройки JSON.NET, там включен camelCase
</blockquote></details>

Здесь именно значение

---

*[VStore/Objects/ObjectManagementService.cs, line 46 at r4](https://reviewable.io:443/reviews/2gis/nuclear-vstore/2#-KZ0VIXEZlYOK5ykkd0a:-KZ0eGWS34pm1teHM9f-:bgltx0j) ([raw file](https://github.com/2gis/nuclear-vstore/blob/4700ab6c424cfb3894ecbbbbe343c4e535c35147/VStore/Objects/ObjectManagementService.cs#L46)):*
<details><summary><i>Previously, denisivan0v (Denis Ivanov) wrote…</i></summary><blockquote>

> ValidationRule

В будущем наверно стоит подумать об объектной модели валидации, где логику инкапсулировать в классах, а не процедурах
</blockquote></details>

Согласен

---

*[VStore/Objects/ObjectManagementService.cs, line 92 at r4](https://reviewable.io:443/reviews/2gis/nuclear-vstore/2#-KZ0MhSz4T5axjHWtoQB:-KZ0dG78MPq_szBGGWc9:b-896fix) ([raw file](https://github.com/2gis/nuclear-vstore/blob/4700ab6c424cfb3894ecbbbbe343c4e535c35147/VStore/Objects/ObjectManagementService.cs#L92)):*
<details><summary><i>Previously, denisivan0v (Denis Ivanov) wrote…</i></summary><blockquote>

> IsFormatted

FasComment не может быть форматированным
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/2gis/nuclear-vstore/2)*

**Comment by xakep139** [comment]
Review status: 31 of 66 files reviewed at latest revision, 6 unresolved discussions.

---

*[VStore/Descriptors/Objects/DateElementValue.cs, line 5 at r4](https://reviewable.io:443/reviews/2gis/nuclear-vstore/2#-KZ0KdXzoobhPKV7Zhw-:-KZ13imMIweAC-cVWMLz:b-896fix) ([raw file](https://github.com/2gis/nuclear-vstore/blob/4700ab6c424cfb3894ecbbbbe343c4e535c35147/VStore/Descriptors/Objects/DateElementValue.cs#L5)):*
<details><summary><i>Previously, denisivan0v (Denis Ivanov) wrote…</i></summary><blockquote>

Тут есть противоречие. Raw исключает BeginDate и EndDate, и наоборот.  Тип `DateElementValue` используется для сериализации, поэтому BeginDate и EndDate окажутся в формате.

Давай либо 
- в Raw хранить интервал в каком-то формате, например `2016-12-01T00:00:00.000Z;2016-12-31T23:59:59.000Z`, и заведем специальный класс, который будет уметь парсить Raw и который мы будем использовать в проверках
- либо вообще откажемся от Raw в этом случае
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/2gis/nuclear-vstore/2)*

**Comment by denisivan0v** [comment]
Reviewed 5 of 73 files at r2, 44 of 51 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/2gis/nuclear-vstore/2)*

**Pull request merged**